flax/nnx/summary.py

```diff
@@ -138,80 +141,71 @@ def __repr__(self):
-    )
-    rows.append(call_info)
-    out = method(obj, *args, **kwargs)
-    call_info.outputs = jax.tree.map(_to_dummy_array, out)
-    return out
-
-  return method_wrapper
-
-
-def _call_obj(
-  obj,
-  input_args: tuple[tp.Any, ...],
-  input_kwargs: dict[str, tp.Any],
-  *,
-  call_method: str,
-  rows: list[CallInfo],
-  node_stats: NodeStats,
-  object_types: set[type],
-):
-  original_methods: dict[type, dict[str, tp.Callable]] = {}
-  try:
-    for obj_type in object_types:
-      methods: dict[str, tp.Callable] = {}
-      for name, top_method in inspect.getmembers(obj_type, inspect.isfunction):
-        if not name.startswith('_') or name == '__call__':
-          methods[name] = top_method
-          method_wrapper = get_method_wrapper(rows, node_stats, top_method)
-          setattr(obj_type, name, method_wrapper)
-
-      original_methods[obj_type] = methods
-
-    top_method = getattr(type(obj), call_method)
-    top_method(obj, *input_args, **input_kwargs)
-  finally:
-    for obj_type, methods in original_methods.items():
-      for name, top_method in methods.items():  # type: ignore[assignment]
-        setattr(obj_type, name, top_method)
+def _get_call_info(jitted, method_name, node_stats, obj, compute_flops: bool, *args, **kwargs):
+  e = jitted.lower(obj, *args, **kwargs)
+  flops = _get_flops(e) if compute_flops else None
+  outputs = e.lowered.out_info[2]
+  output_repr = jax.tree.map(_to_dummy_array, outputs)
+  input_args_info, input_kwargs_info = jax.tree.map(
+    _to_dummy_array, (args, kwargs)
+  )
+  object_id: int = getattr(obj, '_nnx_tabulate_id')
+  node_info = node_stats[object_id]
+  assert node_info is not None
+  path = node_info.path
+  if method_name != '__call__':
+    path = (*path, method_name)
+  return CallInfo(
+    object_id=object_id,
+    type=type(obj),
+    path=path,
+    input_args=input_args_info,
+    input_kwargs=input_kwargs_info,
+    outputs=output_repr,
+    flops=flops
+  )
 
 
 def filter_rng_streams(row: CallInfo):
   return not issubclass(row.type, nnx.RngStream)
 
+def create_obj_env(object_types):
+  "Turn a set of object types into a dictionary mapping (type, method name) pairs to methods"
+  result = {}
+  for obj_type in object_types:
+    for name, top_method in inspect.getmembers(obj_type, inspect.isfunction):
+      if not name.startswith('_') or name == '__call__':
+        result[(obj_type, name)] = top_method
+  return result
+
+def argsave(tracer_args, f):
+  "Wrap a function to save its arguments"
+  n = f.__name__
+  @wraps(f)
+  def wrapper(obj, *args, **kwargs):
+      tracer_args.append((obj, n, args, kwargs))
+      return f(obj, *args, **kwargs)
+  return wrapper
+
+def _overwrite_methods(env):
+  "Overwrite methods with functions from an environment"
+  for (obj_type, name), f in env.items():
+    setattr(obj_type, name, f)
+
+def _get_flops(e) -> int:
+  cost = e.cost_analysis()
+  return 0 if cost is None or 'flops' not in cost else int(cost['flops'])
```

Comment:
```suggestion
  cost = e.cost_analysis()
  if cost is None:
    e = e.compile()
    cost = e.cost_analysis()  
  return 0 if cost is None or 'flops' not in cost else int(cost['flops'])
```

Reply:
When would it be the case that a cost analysis would only be available for a compiled version (rather than a lowered one)?

Reply:
Ah- for cuda! Got it.